python/nutpie/compile_pymc.py

```diff
@@ -258,22 +258,149 @@ def compile_pymc_model(model: "pm.Model", **kwargs) -> CompiledPyMCModel:
                 names.append(f"{base}_{'.'.join(str(i) for i in idx)}")
     coords["unconstrained_parameter"] = pd.Index(names)
 
-    return CompiledPyMCModel(
-        _n_dim=n_dim,
-        dims=model.named_vars_to_dims,
-        _coords=coords,
-        _shapes={name: tuple(shape) for name, _, shape in zip(*shape_info)},
-        compiled_logp_func=logp_numba,
-        compiled_expand_func=expand_numba,
+    dims = model.named_vars_to_dims
+    return dims, coords
+
+
+def _compile_pymc_model_jax(model, *, gradient_backend=None, **kwargs):
```

Comment:
why is this jax specific (except for using that gradient if is indeed jax), seems like it would work with the c-backend (or any other) just fine.

Reply:
Yes, with the pytensor gradient backend it should already work with the c backend or the pytorch backends.
I'm not sure if it is a good idea to introduce the c backend as an option here, if we want to deprecate that soon anyway?

As for pytorch, I think we can add that in a separate PR (also needs changes to the optional dependencies to do properly). We can also add an explicit pytorch (or onnyx) backend in the rust code though, which shouldn't have those issues with the GIL.

Reply:
pytensor has no options to use the backend's autodiff machinery instead of its own, so if we want to do that, we need to bypass pytensor and access the generated function. In the PR we also bypass the shared-variable machinery of pytensor, so we access the compiled jax functions directly even if we use pytensor gradients.

This doesn't necessarily have to be like this, but I think I still prefer doing it that way, because the `with_data` machinery won't work otherwise, and we add more python overhead if we go through the pytensor function.
I'd also rather specify the optional dependencies correctly, and make sure things are tested, which is easier if we do this on a backend-by-backend basis.

Reply:
OK, I see, you're not calling the PyTensor function but the underlying JAX directly (through Python). That all makes sense then.